[101129996] Install cloudfoundry nats service on AWS

keymon opened the pull request:
[Create NATS instance in CF environment on AWS](https://www.pivotaltracker.com/n/projects/1275640/stories/101129996)

> Note: This PR is a continuation of #3, which was closed, and it includes additional changes from that (closed) PR. 
> The most important changes:
> - stop using EIP for microbosh (unnecessary)
> - adjust instances (disk sizes, instance types)
> - ~~create subnet for NATS~~ (not used at the end)

Using the existing microbosh and bastion host deployed using terraform and bosh-init, this story extends the CF landscape to include define the first steps and manifests to deploy the NATS component. 

We agreed to build out the CF landscape one component at a time and this is the first of these stories. 

Initially the story was meant for GCE+AWS, but it got split and we will focus only on AWS.

**Acceptance Criteria**
**Given** that there is a script for deploying a CF build in AWS
**When** I run the script 
**Then** an instance of NATS is deployed in the environment
## Scope of the PR

This story should be reviewed in this context
- I want to have a initial basic CF manifest which includes:
  - a job to deploy the nats service
  - all the required elements: pools, compile pools, network definitions, etc.
- I want to be able to upload the manifest on the existing microbosh.
- I want to be able to do a `bosh deployment` and have a running instance of nats.
- ~~I want to have the necessary security groups that allow everything to work.~~

> Note: We discarded implement security groups due limitations on BOSH CPI. see git comments.
## How to review the card.

We carry on implementing tasks on `Makefile` to deploy the environment. To provision the environment simply run:

```
make aws DEPLOY_ENV=<name>
```

If it finishes successfully, you can try to connect to nats with the nats client from the bastion host

```
# Install nats client
sudo gem install nats --no-ri --no-rdoc

# Listen for nats messages in background
nats-sub '>' -s nats://nats:nats-password@10.0.0.10:4222 & 

# Send a message
nats-pub foo 'Hello World!' -s nats://nats:nats-password@10.0.0.10:4222 
```
## Detailed PR description

The detailed execution of the deployment will be:
- `apply-aws`: Apply terraform to create all the AWS required elements, including:
  - networks and security groups.
  - create the bastion host
  - Render the templated manifests. We use a local provisioner to create a local file and we upload it via SSH. 
- `prepare-provision`: Upload the manifests to the bastion
- `provision`: Run the script `provision.sh` which will deploy microbosh and CF:
  1. Deploy microbosh using `bosh-init`
  2. Deploy CF (and nats) using `bosh deploy`

The manifest file is based on [this example manifest from `cf-release`](https://github.com/cloudfoundry/cf-release/blob/master/example_manifests/minimal-aws.yml).

mtekel committed:
microbosh: change instance to t2.medium

Change microbosh instance from m3.xlarge to t2.medium. Microbosh will sit idle most of its life and the only time it needs CPU is on the initial deployment and when doing changes to CF. T2 instance type is well suited to this bursty behaviour. 4GB memory is enough. m3.xlarge is unnecessary overkill, costing 5x more.

mtekel committed:
bastion host: increase disk size

Set disk size to 100GB for bastion host. The default image only has 7GB free, which is not enough to download all the stemcells, releases and prepare tarball. We might also want to keep 2 releases on the bastion host in case we want to do rollback.

mtekel committed:
firewall: only local network for bastion to bosh

There's no need to access microbosh on it's external IP when it resides on the same network as bastion host. Only allow bastion to microbosh communication to happen on local network. No other communication to microbosh is allowed (not even from office: if you want to manipulate microbosh, do it via bastion host).

mtekel committed:
use local IP for microbosh

Define microbosh static IP as 10.0.0.6, which is traditional microbosh IP. We want to have same IP in all our clouds. We don't want to access microbosh on EIP, just on the internal IP. Use the aws_static_ip everywhere microbosh IP is used.

mtekel committed:
aws: define CF subnets

Define 2 subnets for CF:
* 10.0.1/2.0/24 for core CF components
* 10.0.11/12.0/24 for app related components (routers, DEAs)

mtekel committed:
bastion host: use fixed IP

Use 10.0.0.4 fixed IP for bastion host. As we might want bosh to deploy something in the spare range of this subnet (say from 10.0.0.10 to 10.0.0.250), we fix the IP of bastion host to 10.0.0.4, the 1st available IP in AWS subnets (.1 is gateway, .2 DNS and .3 is also reserved). Specify bastion host VPC security group explicitly, using correct terraform argument.

mtekel committed:
provision script: add bosh cli, v215 release

Add more actions to the provision script:
* install bosh cli (from rubygems)
* download (from internet) and upload (to bosh) stemcell
* download (from internet) and upload (to bosh) v215 release
* add comments about known issues and pending work

This is work in progress script. The TODO and FIXME issues will be fixed in separate stories. Script will be cleaned up and beautified then.

mtekel committed:
networks: rename bastion routing table to internet

This routing table gives direct access to the internet. Rename it to 'internet' to make more obvious what it does. This way we can re-use it for multiple subnets and still easily keep track of what's going on.

mtekel committed:
networks: add CF core subnet

Create cf core subnet (just 1 zone for now) which we have defined in variables.

keymon committed:
Fix error handling and optimise `aws/provision.sh`

Check if some resources are already present, update
the cf-release, etc.

keymon committed:
Split the security groups for microbosh and vas

Microbosh security group and rules is different than the 
default security group for the VMs provision by Bosh.

Microbosh must allow access from bastion for SSH, nats, mbus, director api, etc, so that 
`bosh-init` can deploy the agent and later we can use the `bosh_cli`. Additionally ot must allow access to the `bosh registry` and the `nats` server to any provision VM.

The provision VMs should allow access to SSH from the bastion host and the director (microbosh).

keymon committed:
Add AWS security group for nats service

Allow access to nats port (4222) from any provision vm. Also
from the bastion to help troubleshooting.

keymon committed:
Add first manifest for cloud foundry on AWS

This manifest is based on the [minimal AWS manifest example](https://github.com/cloudfoundry/cf-release/blob/master/example_manifests/minimal-aws.yml) from cf-release.

Only includes the `nats` job, in the static IP 10.0.0.10.

Defines: 
 - a network with some predefined static IPs
 - compilation pools and small vm pool
 - some defaults for the cloud properties.

Important: The compilation job needs ephemeral disk. Currently setup using a
global `cloud_properties.ephemeral_disk` entry. but not all VMs will need this 
disk, we should change it later.

RichardKnop committed:
Fixed a syntax error

keymon committed:
Break cycle dependency on security groups

RichardKnop committed:
Fixed stemcell in CF manifest

RichardKnop committed:
Moved ephemeral disk definition to compilation pool

Moved ephemeral disk and availability zone definitions to compilation
pool, and also availability zone to resource pool.

Increased number of compilation instances and the type from t2.medium to
c3.large.

keymon committed:
Add ephemeral disk to AWS small vm resource_pool

All the provision VMs require a ephemeral disk. If not provided the bosh agent
will fail, restart and not call back to the BOSH nats service.

Likely we will need to move this as a default property in the future.

keymon committed:
Move AWS manifest templates to dedicated file

As we add more manifests it is not so related to the bastion host.

keymon committed:
Move upload of manifest out of instance provision

The provisioners associated to the instances will only update
the files if the instance is reprovision. This means that
we will need to reprovision the bastion host every time we want
to change the template.

To avoid this, we will create the files using a `local-exec` provisioner
in the terraform template and upload them using ssh in the Makefile.

One restriction of this is that the template must not include a single
quote, as it might interfere with the local-exec of the template.

RichardKnop committed:
Moved security groups from job to resource pools.

It would seem, according to docs, that cloud_properties can only be
defined on resource pools, so we will have to create distinct resource
pools for NATS and other components in order to be able to have granular
control over security groups per component.

See: https://bosh.io/docs/deployment-manifest.html

keymon committed:
Downloading/uploading a specific HVM stemcell for VMs

keymon committed:
Implement logic run CF deployment from Make on AWS

keymon committed:
Do not automatically associate public IP on AWS

We must not automatically associate public IPs to all the instances in
the bastion subnet. Instead, we will explicitelly do so as needed,
for instance in this case on the bastion host.

keymon committed:
Remove unused AWS subnet cf_core

Currently we use the bastion subnet for all the components and the
cloudfoundry installation itself. We might implement further separation
in the future.

RichardKnop committed:
Can't have spaces in manifest name so removed them.

keymon committed:
Revert "Implement logic run CF deployment from Make on AWS"

This reverts commit 3e45b359f6241d43002ded850a4e95fc279f48d8.

Conflicts:
	aws/cf-manifest.yml.tpl

keymon committed:
Update the BOSH UUID in the provision.sh file

keymon committed:
Increase AWS CF release to v215

keymon committed:
Remove nats security group on AWS CF definition

AWS BOSH CPI only allows define the security group per network,
not even by resource pool. 

That means define one network per service, which will be
quite tedious. Because that, for the time being, we will drop
the approach of different security groups per service and use
only one common security group.

keymon committed:
Remove unused nats AWS security group definition 

As we are not going to use it for the time being.

keymon committed:
Open all internal traffic in default AWS sg

We will use a common security group on all the vms provision
by bosh, and allow all traffic from the internal network.

keymon committed:
Force tty when running provision.ssh via ssh

When we run ssh from `Makefile` we must pass the option `-t` to
force allocation of a tty. Without it, interactive prompts (e.g. bosh) will
fail.

keymon committed:
Remove unnecessary stemcell upload in AWS

We don't need to upload two stemcells, only the hvm one

jimconner commented on aws/manifest.yml.tpl:
```diff
@@ -55,10 +57,8 @@ jobs:
+    static_ips: [${aws_static_ip}]
     default: [dns, gateway]
-  - name: public
-    static_ips: [${aws_static_ip}] # <--- Replace with Elastic IP
```
This static IP is required by the microbosh so that it can communicate with AWS in order to spin up new VMs, upload stemcells etc.

jimconner commented on aws/network.tf:
```diff
@@ -1,7 +1,3 @@
-resource "aws_eip" "bosh" {
-  vpc = true
-}
-
```
This static IP is required by the microbosh so that it can communicate with AWS in order to spin up new VMs, upload stemcells etc.

jimconner commented on aws/outputs.tf:
```diff
@@ -1,7 +1,3 @@
-
-output "bosh_ip" {
-	value = "${aws_eip.bosh.public_ip}"
-}
```
This static IP is required by the microbosh so that it can communicate with AWS in order to spin up new VMs, upload stemcells etc.

keymon replied:
Fixed by adding the EIP

keymon committed:
Use public EIP for AWS microbosh to access inet

The microbosh machine needs to be able to reach the internet, to
download binaries and connect to the AWS API.

In order to do that, we can either setup a NAT host to route (which is a SPOF)
or assign a public IP to the host. We apply here the second approach.

The security groups associated to the director will block any incomming
traffic.

This commit partially reverts d2986a5ff9d29fdaa4b8b5944431c69b74651726

jimconner commented on aws/provision.sh:
```diff
@@ -14,6 +17,53 @@ chmod 400 ~/.ssh/id_rsa.pub
+cd ~
+sed -i "s/BOSH_UUID/$(bosh status --uuid)/" cf-manifest.yml
+bosh deployment cf-manifest.yml
+time bosh deploy
```
If you do a 'bosh -n deploy' here instead of just 'bosh deploy' then it will be in non-interactive mode and will not prompt the user to answer 'yes' to deploying.

keymon replied:
I will add the option. Thx for the finding!

jimconner commented:
I have tested on AWS and confirmed deployment of a nats machine, and the ability to send and receive messages over the NATS message bus.

keymon committed:
Add -n option to bosh deploy to proceed w/o confirmation

RichardKnop commented:
Works now. Thanks @keymon @jimconner for fixing these two issues :+1:

actionjack commented:
Wow that was a bit of a mission! Merging...
